packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -255,110 +298,140 @@ export class ChainService implements ChainServiceInterface {
+      ...(this.channelToSubscribers.get(channelId) ?? []),
+      subscriber,
+    ]);
+
+    assetHolders.map(assetHolder => {
+      this.setUpAssetHolderListener(assetHolder);
       const contract = this.getOrAddContractMapping(assetHolder);
       if (!contract) throw new Error('The addressToContract mapping should contain the contract');
-      const currentHolding = from(this.getInitialHoldings(contract, channelId));
-
-      const subscription = concat<ContractEvent>(
-        currentHolding,
-        obs.pipe(filter(event => event.channelId === channelId))
-      ).subscribe({
-        next: async event => {
-          switch (event.type) {
-            case Deposited:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed Deposited event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.holdingUpdated(event);
-              break;
-            case AssetTransferred:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed AssetTransferred event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.assetTransferred(event);
-              break;
-            default:
-              throw new Error('Unexpected event from contract observable');
-          }
-        },
-      });
-      const subscriptions = this.channelToSubscription.get(channelId) ?? [];
-      this.channelToSubscription.set(channelId, [...subscriptions, subscription]);
+      // Fetch the current contract holding, and emit as an event
+      this.getInitialHoldings(contract, channelId).then(initialHoldings =>
+        subscriber.holdingUpdated(initialHoldings)
+      );
     });
   }
 
-  /** Implementation note:
-   *  The following is a simplified API that assumes a single registerChannel call per channel.
-   *  If we would like to allow for multiple registrations per channel, registerChannel should return a registration ID.
-   *  unregisterChannel will require the registration ID as a parameter.
-   */
   unregisterChannel(channelId: Bytes32): void {
-    const subscriptions = this.channelToSubscription.get(channelId);
-    if (subscriptions?.length !== 1) {
-      throw new Error(
-        'Unregister channel implementation only works when there is one subscriber per channel'
-      );
+    this.channelToSubscribers.set(channelId, []);
+  }
+
+  private async whenFinalized(channelId: Bytes32, finalizesAtS: number, tryNum: number) {
+    if (tryNum > 4) return;
+
+    const nowMs = Date.now();
+    const finalizesAtMs = finalizesAtS * 1_000;
+    if (nowMs < finalizesAtMs) {
+      setTimeout(() => this.whenFinalized(channelId, finalizesAtS, 0), nowMs - finalizesAtMs);
+      return;
     }
```

Comment:
It seems to me like we can leverage the fact that the `ethers` library exposes `provider.on('block', block => ...)` which emits each time a new block is produced. Instead of relying on `Date.now()` we could rely on `block.timestamp`

Reply:
I think that's an interesting point. But what if a challenge is 2 weeks? Do you still want to fire a callback on every block?

Reply:
That approach is probably ok as long as there is one global callback instead of one callback per channel.

Reply:
Considering that the `ChainService` is already inspecting every block (e.g., for `Deposited` events), if it's just doing a single additional comparison operation, I don't think it's adding any meaningful overhead in practice. Arguably the code simplification benefits might outweigh the extra check

Reply:
Good point!

Reply:
Addressed in 0d194d4b9.